### PR TITLE
Implementation of exti on esp8266

### PR DESCRIPTION
**Pull request by wuwx**
:)

### CODE REVIEW

File: src/drivers/ports/esp/exti_port.i [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+static int exti_port_module_init()
+{
+    _xt_isr_attach(ESP8266_IRQ_NUM_GPIO, isr, NULL);
+    _xt_isr_unmask(1<<ESP8266_IRQ_NUM_GPIO);
+    return (0);
+}
+
+int exti_port_init(struct exti_driver_t *drv,
```
Comment (eerimoq) [inline src/drivers/ports/esp/exti_port.i]: Unused function, remove.

File: src/drivers/ports/esp/exti_port.i [hunk 2]
```diff
@@ -0,0 +1,81 @@
+/**
```
Comment (eerimoq) [inline src/drivers/ports/esp/exti_port.i]: A general comment: Always add the suffix ``_p`` to pointers to make the code easier to read.

File: src/drivers/ports/esp/exti_port.i [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+int exti_port_init(struct exti_driver_t *drv,
+                   struct exti_device_t *dev,
+                   void (*on_interrupt)(struct exti_driver_t *drv))
+{
+    return (0);
+}
+
+int exti_port_start(struct exti_driver_t *drv)
```
Comment (eerimoq) [inline src/drivers/ports/esp/exti_port.i]: Replace ``drv`` with ``self_p``

File: src/drivers/ports/esp/exti_port.i [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    return (0);
+}
+
+int exti_port_stop(struct exti_driver_t *drv)
+{
+    gpio_pin_intr_state_set(drv->dev_p->pin_p->id, 0);
+    return (0);
+}
+
+static int exti_port_clear(struct exti_driver_t *self_p)
```
Comment (eerimoq) [inline src/drivers/ports/esp/exti_port.i]: The driver should only clear the interrupt flag bit for its pin, not all pins.

File: src/drivers/ports/esp/exti_port.i [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+    pin_reg = ESP8266_GPIO->CONF[i];
+    pin_reg &= (~ESP8266_GPIO_CONF_INT_TYPE_MASK);
+    pin_reg |= (intr_state << ESP8266_GPIO_CONF_INT_TYPE_POS);
+    ESP8266_GPIO->CONF[i] = pin_reg;
+
+    portEXIT_CRITICAL();
+}
+
+extern volatile int count;
```
Comment (eerimoq) [inline src/drivers/ports/esp/exti_port.i]: Remove unused variable ``count``.

**Comment by wuwx** [comment]
i made a key press, and change the count value, in other thrd output the count value
i don't know how to test exti ...

and i found i can't use std_printf in the callback , if use, the system die...

**Comment by wuwx** [comment]
i will change the code and give a new pull request

**Comment by eerimoq** [comment]
Yeah the std_printf() function can not be used in interrupts. only functions with the suffix ``_isr`` should be called from and isr. The pin module is an exception to this. It's okay to read and write to pins from an interrupt handler.

**Comment by eerimoq** [comment]
Thanks, I'll accept the pull request.

**Comment by wuwx** [comment]
also thank you
and if i want print some message if user press a key(pin)
how to print a message to uart

**Comment by eerimoq** [comment]
You can use the [event module](http://simba-os.readthedocs.io/en/latest/library-reference/sync/event.html) to signal an event from the isr to a thread and print from the thread. Usually nothing but signalling a thread should be done from an interrupt handler as interrupts handlers should return as quickly as possible.

**Pull request merged**